File: native/nan_breaker.c
/* Babel: C — NaN circuit breaker for training step loops. */
#include "nan_breaker.h"

void nan_breaker_init(NanBreaker *b, int trip_after) {
    b->trip_after = trip_after > 0 ? trip_after : 3;
    b->consecutive_bad = 0;
    b->open = 0;
}

int nan_breaker_observe(NanBreaker *b, int finite) {
    if (b->open) return 1;
    if (finite) b->consecutive_bad = 0;
    else {
        b->consecutive_bad++;
        if (b->consecutive_bad >= b->trip_after) b->open = 1;
    }
    return b->open;
}

int nan_breaker_allow(const NanBreaker *b) { return !b->open; }

void nan_breaker_reset(NanBreaker *b) {
    b->open = 0;
    b->consecutive_bad = 0;
}

File: src/nan_breaker.py
"""NaN circuit breaker — trip on consecutive non-finite steps."""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class BreakerState(str, Enum):
    CLOSED = "CLOSED"  # normal
    OPEN = "OPEN"      # tripped


@dataclass
class NanCircuitBreaker:
    trip_after: int = 3
    consecutive_bad: int = 0
    state: BreakerState = BreakerState.CLOSED

    def observe(self, finite: bool) -> BreakerState:
        if self.state is BreakerState.OPEN:
            return self.state
        if finite:
            self.consecutive_bad = 0
        else:
            self.consecutive_bad += 1
            if self.consecutive_bad >= self.trip_after:
                self.state = BreakerState.OPEN
        return self.state

    def allow_step(self) -> bool:
        return self.state is BreakerState.CLOSED

    def reset(self, token: str) -> None:
        if not token.startswith("reset:"):
            raise ValueError("BAD_RESET")
        self.state = BreakerState.CLOSED
        self.consecutive_bad = 0
